src/neuron/container/soa_container.hpp

```diff
@@ -83,6 +84,16 @@ struct has_num_variables<T, std::void_t<decltype(std::declval<T>().num_variables
 template <typename T>
 inline constexpr bool has_num_variables_v = has_num_variables<T>::value;
 
+template <class T>
+auto get_num_variables(T const& t) -> std::enable_if_t<has_num_variables_v<T>, size_t> {
+    return t.num_variables();
+}
+
+template <class T>
+auto get_num_variables(T const& t) -> std::enable_if_t<!has_num_variables_v<T>, size_t> {
+    return 1ul;
+}
```

Comment:
Completely optional, but
```suggestion
template <class T>
std::size_t get_num_variables(T const& t) {
  if constexpr (has_num_variables_v<T>) {
    return t.num_variables();
  } else {
    return 1;
  }
}
```
might be friendlier 🤷

Reply:
Consistency with the lines immediate prior to this one would suggest not. Nevertheless, changed.

Reply:
If you mean `get_array_dimension`, I think that one has to be a little more complex to handle passing optional arguments to the optional method.